py/kubeflow/kubeflow/ci/update_jupyter_web_app.py

```diff
@@ -0,0 +1,269 @@
+        isn't present.
+    """
+    for r in repo.remotes:
+      for u in r.urls:
+        if remote_url == u:
+          return r
+
+    return None
+
+  def all(self, build_project, registry_project, remote_fork,
+          add_github_host=False): # pylint: disable=too-many-statements,too-many-branches
+    """Build the latest image and update the prototype.
+
+    Args:
+      build_project: GCP project used to build the image.
+      registry_project: GCP project used to host the image.
+      remote_fork: Url of the remote fork.
+        The remote fork used to create the PR;
+         e.g. git@github.com:jlewi/kubeflow.git. currently only ssh is
+         supported.
+      add_github_host: If true will add the github ssh host to known ssh hosts.
+    """
+    repo = git.Repo(self._root_dir())
+    util.maybe_activate_service_account()
+    last_commit = self.last_commit
+
+    # Ensure github.com is in the known hosts
+    if add_github_host:
+      output = util.run(["ssh-keyscan", "github.com"])
+      with open(os.path.join(os.getenv("HOME"), ".ssh", "known_hosts"),
+                mode='a') as hf:
+        hf.write(output)
+
+    if not remote_fork.startswith("git@github.com"):
+      raise ValueError("Remote fork currently only supports ssh")
+
+    remote_repo = self._find_remote_repo(repo, remote_fork)
+
+    if not remote_repo:
+      fork_name = remote_fork.split(":", 1)[-1].split("/", 1)[0]
+      logging.info("Adding remote %s=%s", fork_name, remote_fork)
+      remote_repo = repo.create_remote(fork_name, remote_fork)
+
+    logging.info("Last change to components-jupyter-web-app was %s", last_commit)
+
+    base = "gcr.io/{0}/jupyter-web-app".format(registry_project)
+
+    # Check if there is already an image tagged with this commit.
+    image = base + ":" + self.last_commit
+    transport = transport_pool.Http(httplib2.Http)
+    src = docker_name.from_string(image)
+    creds = docker_creds.DefaultKeychain.Resolve(src)
+
+    image_exists = False
+    try:
+      with v2_2_image.FromRegistry(src, creds, transport) as src_image:
+        logging.info("Image %s exists; digest: %s", image,
+                     src_image.digest())
+        image_exists = True
+    except docker_http.V2DiagnosticException as e:
+      if e.status == 404:
+        logging.info("%s doesn't exist", image)
+      else:
+        raise
+
+    if not image_exists:
+      logging.info("Building the image")
+      image = self.build_image(build_project, registry_project)
+    else:
+      logging.info("Image %s already exists", image)
+
+    # TODO(jlewi):We should check what the current image and not update it
+    # if its the existing image
+    prototype_file = self.update_prototype(image)
+
+    branch_name = "update_jupyter_{0}".format(last_commit)
+
+    if repo.active_branch.name != branch_name:
+      logging.info("Creating branch %s", branch_name)
+      util.run(["git", "checkout", "-b", branch_name], cwd=self._root_dir())
```

Comment:
The branch name could already exist if this script is run multiple times, for the same commit hash.

Reply:
Added a check.